charts/HTMLTimeline.tsx

```diff
@@ -52,102 +58,91 @@ export class Timeline extends React.Component<TimelineProps> {
         }
-
-        runInAction(() => {
-            this.startYearInput = props.startYear
-            this.endYearInput = props.endYear
-        })
     }
 
     componentDidUpdate() {
         const { isPlaying, isDragging } = this
         if (!isPlaying && !isDragging) {
             runInAction(() => {
-                this.startYearInput = this.props.startYear
-                this.endYearInput = this.props.endYear
+                this.startYearRaw = this.props.startYear
+                this.endYearRaw = this.props.endYear
             })
         }
     }
 
-    @computed get years(): number[] {
+    @computed get years(): Time[] {
         return this.props.years
     }
 
-    @computed get minYear(): number {
-        // This cast is necessary because `years` might be empty. Should deal
-        // with an empty years array more gracefully -@jasoncrawford 2019-12-17
-        return first(this.props.years) as number
+    @computed get minYear(): Time {
+        return first(this.props.years) ?? DEFAULT_MIN_YEAR
     }
 
-    @computed get maxYear(): number {
-        // This cast is necessary because `years` might be empty. Should deal
-        // with an empty years array more gracefully -@jasoncrawford 2019-12-17
-        return last(this.props.years) as number
+    @computed get maxYear(): Time {
+        return last(this.props.years) ?? DEFAULT_MAX_YEAR
     }
 
-    // Sanity check the input
-    @computed get startYear(): number {
-        const { startYearInput, endYearInput, minYear, maxYear } = this
-        return Math.min(
-            maxYear,
-            Math.max(minYear, Math.min(startYearInput, endYearInput))
-        )
+    @computed get timeDomain(): [Time, Time] {
+        return [this.minYear, this.maxYear]
     }
 
-    // Closest year to the input start year
-    // e.g. 1954 => 1955
-    @computed get roundedStartYear(): number {
-        const { years, startYear } = this
-        return sortBy(years, year => Math.abs(year - startYear))[0]
+    getClampedYear(inputYear: Time): Time {
+        const { minYear, maxYear } = this
+        return Math.min(maxYear, Math.max(minYear, inputYear))
     }
 
-    // Previous year from the input start year
-    // e.g. 1954 => 1950
-    @computed get targetStartYear(): number {
-        const { years, startYear } = this
-        return find(
-            sortBy(years, year => Math.abs(year - startYear)),
-            year => year <= startYear
-        ) as number
+    getYearUI(bound: TimeBound): Time {
+        if (isUnboundedLeft(bound)) return this.minYear
+        if (isUnboundedRight(bound)) return this.maxYear
+        return bound
     }
 
-    @computed get endYear(): number {
-        const { startYearInput, endYearInput, minYear, maxYear } = this
-        return Math.min(
-            maxYear,
-            Math.max(minYear, Math.max(startYearInput, endYearInput))
-        )
+    getYearRound(bound: TimeBound, defaultValue: TimeBound): TimeBound {
```

Comment:
Well, it's not really rounding...

Reply:
Yes, I had actually renamed this method to something else. And then came back to Round because it felt most appropriate.

The timeline feeds decimals into this function and it returns the closest value in an array... not quite round, but not sure what to call it.

Reply:
Ah alright, that makes more sense. I guess I was confused because I thought `TimeBound` can only ever be an integer (or infinity), but right, it can be everything.
But even then, it isn't intuitive that this can still return +-Infinity. Judging from the name one would think this would find some year from the timeline.

Reply:
Also renamed this to `{start,end}YearClosest`, feels righter...